### PR TITLE
Reusable docstring parsing and more metadata checks 

### PR DESCRIPTION
**Pull request by timburke**
This PR adds three major things:

1. It adds support for parsing lists that are tagged as annotated parameters using the type code `list(x)` where x is another supported type like `list(string)` or `list(integer)`.  The list is parsed by passing it to `ast.literal_eval` which will parse any valid python list like:
```
["a", "b", "c"]
```
You still need to make sure it is treated as a single parameter with appropriate quoting around the whole thing if it has spaces and you need to make sure that any strings inside the array are quoted as required for python strings.  Since each individual item is passed to a string to type converter you can pass strings instead of integers that will be parsed or you can pass integers directly.  If you pass bools, they must be python bools (`True`, `False`).  If you instead quote them as strings, you don't need to capitalize them.

Examples:

```
> iotile some_function --list "[1, 2, 3]"
> iotile other_functiion --arg2 '["a", "b", "c"]'
> iotile third_function --bool_list [True,False,False]
```

2. It adds experimental support for extracting structured information from doctstrings.  This is used in `iotile-analytics` for showing help information on required parameters that users can set when working with report generation.  

3.  It increases slightly the capabilities of `AnnotatedMetadata` to allow checking if a given set of function parameters are correct for calling a function.   This is also used inside `iotile-analytics`.

## User Visible Changes

1. The only major change is the ability to now specify things like `list(bool)` in an `@param` decorator.

### CODE REVIEW

File: typedargs/doc_parser.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+            paragraphs.append(cls._join_paragraph(curr_para, leading_blanks, trailing_blanks))
+
+        return paragraphs
+
+    def wrap_and_format(self, width=None, include_params=False, include_return=False, excluded_params=None):
+        """Wrap, format and print this docstring for a specific width.
+
+        Args:
+            width (int): The number of characters per line.  If set to None
+                this will be inferred from the terminal width and defualt
```
Comment (haroal) [inline typedargs/doc_parser.py]: default*

File: typedargs/doc_parser.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+                    curr_para = [line.lstrip()]
+            else:
+                if len(line) != 0:
+                    curr_para.append(line)
+                else:
+                    paragraphs.append(cls._join_paragraph(curr_para, leading_blanks, trailing_blanks))
+                    curr_para = []
+
+        # Finish the last paragraph if ther is one
```
Comment (haroal) [inline typedargs/doc_parser.py]: there*

File: typedargs/doc_parser.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        raise ValidationError("Invalid parameter type string not enclosed in ( ) characters", param_string=param_def, type_string=param_type)
+
+    param_type = param_type[1:-1]
+    return param_name, ParameterInfo(param_type, [], desc)
+
+
+def parse_return(return_line, include_desc=False):
+    """Parse a single return statement declaration.
+
+    The valid types of return declarion are a Returns: section heading
```
Comment (haroal) [inline typedargs/doc_parser.py]: declaration*

Comment (timburke) [reply typedargs/doc_parser.py]: resolved. thanks for finding these!

**Pull request merged**